model/host/db.go

```diff
@@ -116,6 +119,77 @@ func AllIdleEphemeral() ([]Host, error) {
 	return Find(query)
 }
 
+// AllHostsSpawnedByTasksToTerminate finds all hosts spawned by tasks that should be terminated.
+func AllHostsSpawnedByTasksToTerminate() ([]Host, error) {
+	catcher := grip.NewBasicCatcher()
+	var hosts []Host
+	timedOutHosts, err := AllHostsSpawnedByTasksTimedOut()
+	hosts = append(hosts, timedOutHosts...)
+	catcher.Add(err)
+
+	taskHosts, err := AllHostsSpawnedByFinishedTasks()
+	hosts = append(hosts, taskHosts...)
+	catcher.Add(err)
+
+	buildHosts, err := AllHostsSpawnedByFinishedBuilds()
+	hosts = append(hosts, buildHosts...)
+	catcher.Add(err)
+
+	if catcher.HasErrors() {
+		return nil, catcher.Resolve()
+	}
+	return hosts, nil
+}
+
+// AllHostsSpawnedByTasksTimedOut finds hosts spawned by tasks that should be terminated because they are past their timeout.
+func AllHostsSpawnedByTasksTimedOut() ([]Host, error) {
+	query := db.Query(bson.M{
+		StatusKey: evergreen.HostRunning,
+		bsonutil.GetDottedKeyName(SpawnOptionsKey, SpawnOptionsTimeoutKey): bson.M{"$lte": time.Now()},
+	})
+	return Find(query)
+}
+
+// AllHostsSpawnedByFinishedTasks finds hosts spawned by tasks that should be terminated because their tasks have finished.
+func AllHostsSpawnedByFinishedTasks() ([]Host, error) {
+	const runningTasks = "running_tasks"
+	pipeline := []bson.M{
+		{"$lookup": bson.M{
+			"from":         task.Collection,
+			"localField":   bsonutil.GetDottedKeyName(SpawnOptionsKey, SpawnOptionsTaskIDKey),
+			"foreignField": task.IdKey,
+			"as":           runningTasks,
+		}},
+		{"$unwind": "$" + runningTasks},
```

Comment:
should we redact the running_tasks field from these documents to avoid bloat back over the wire?